File: mobile/android/remote-core/src/main/kotlin/com/procway/pleiad/remote/DeviceLink.kt
package com.procway.pleiad.remote

// Kotlin port of core/remote/device-link.mjs (docs/remote.md §3.2・§4.4・§7.4): one line from the device through the
// relay to the host, with reconnects and states. Loop-confined; the blocking relay handshake runs on a worker thread.
//
// state: connecting | connected | offline | host-offline | revoked | stopped
// Backoff 0.5 s -> 30 s doubling (±25 % jitter), reset after 10 s connected. Streams never survive a reconnect.

import java.util.concurrent.Executors
import org.json.JSONObject

data class LinkStatus(
    val state: String,
    val since: Long = System.currentTimeMillis(),
    val retryAt: Long? = null,
    val closeCode: Int? = null,
    val goaway: String? = null,
    val reason: String? = null,
    val httpStatus: Int? = null,
    val hostName: String = "",
    val connectedAt: Long? = null,
) {
    fun toJson(): JSONObject = JSONObject().apply {
        put("state", state)
        put("since", since)
        retryAt?.let { put("retryAt", it) }
        closeCode?.let { put("closeCode", it) }
        goaway?.let { put("goaway", it) }
        reason?.let { put("reason", it) }
        put("hostName", hostName)
        connectedAt?.let { put("connectedAt", it) }
    }
}

class LinkUnavailable(val state: String, val closeCode: Int? = null) : Exception("link $state")

data class Backoff(val minMs: Long = 500, val maxMs: Long = 30_000, val stableMs: Long = 10_000)

object LinkRules {
    fun classifyClose(closeCode: Int? = null, goaway: String? = null): String = when {
        goaway == "revoked" || closeCode == 4401 -> "revoked"
        goaway == "shutdown" || closeCode == 4404 || closeCode == 4408 -> "host-offline"
        else -> "offline"
    }
}

class DeviceLink(
    val loop: Loop,
    @Volatile var creds: HostCreds,
    private val keyPair: KeyPair,
    private val app: String = "",
    private val name: String = "",
    private val shell: String = "mobile",
    private val backoff: Backoff = Backoff(),
    private val connectTimeoutMs: Long = 15_000,
    private val channelFactory: ((sendFn: (ByteArray) -> Unit, transport: Transport, hello: JSONObject, buffered: () -> Long) -> Channel)? = null,
    private val log: (String) -> Unit = {},
) {
    companion object {
        /** Blocking handshakes run here (never on the loop). */
        private val workers = Executors.newCachedThreadPool { r -> Thread(r, "pleiad-link").also { it.isDaemon = true } }
    }

    var channel: Channel? = null; private set
    var hostName = ""; private set
    private var attempt = 0
    private var running = false
    private var generation = 0
    private var retryTimer: Cancellable? = null
    private var stableTimer: Cancellable? = null
    private var socket: RelaySocket? = null
    private var connectedAt: Long? = null
    private val statusListeners = mutableListOf<(LinkStatus) -> Unit>()
    var status = LinkStatus("stopped"); private set
    val state get() = status.state

    fun onStatus(fn: (LinkStatus) -> Unit) { statusListeners.add(fn) }

    private fun setStatus(s: LinkStatus) {
        status = s.copy(hostName = hostName, connectedAt = if (s.state == "connected") connectedAt else null)
        for (l in statusListeners.toList()) try { l(status) } catch (_: Exception) {}
    }

    fun start() = loop.exec {
        if (running) return@exec
        running = true
        attempt = 0
        connect()
    }

    fun stop() = loop.exec {
        if (!running && state == "stopped") return@exec
        running = false
        generation++
        retryTimer?.cancel(); retryTimer = null
        stableTimer?.cancel()
        val ch = channel
        channel = null
        ch?.close(ChannelError("closed", "link stopped"))
        socket?.close(1000)
        socket = null
        setStatus(LinkStatus("stopped"))
    }

    /** Reconnect now ("retry", foreground again, or new credentials after re-pairing). */
    fun retryNow(newCreds: HostCreds? = null) = loop.exec {
        if (newCreds != null) creds = newCreds
        if (!running) { running = true; attempt = 0; connect(); return@exec }
        if ((state == "connecting" || state == "connected") && newCreds == null) return@exec
        generation++
        retryTimer?.cancel(); retryTimer = null
        val ch = channel
        channel = null
        ch?.close(ChannelError("closed", "reconnecting"))
        socket?.close(1000)
        socket = null
        attempt = 0
        connect()
    }

    /**
     * The usable channel: at once when connected, after the attempt when connecting (max ms), otherwise fail at once
     * with LinkUnavailable(state). The callback runs on the loop.
     */
    fun ready(ms: Long, cb: (Channel?, LinkUnavailable?) -> Unit) = loop.exec {
        val ch = channel
        if (state == "connected" && ch != null && !ch.closed) return@exec cb(ch, null)
        if (state != "connecting") return@exec cb(null, LinkUnavailable(state, status.closeCode))
        var done = false
        var timer: Cancellable? = null
        lateinit var listener: (LinkStatus) -> Unit
        listener = { s ->
            if (!done && s.state != "connecting") {
                done = true
                statusListeners.remove(listener)
                timer?.cancel()
                val c = channel
                if (s.state == "connected" && c != null) cb(c, null) else cb(null, LinkUnavailable(s.state, s.closeCode))
            }
        }
        statusListeners.add(listener)
        timer = loop.schedule(ms) {
            if (!done) {
                done = true
                statusListeners.remove(listener)
                cb(null, LinkUnavailable("offline"))
            }
        }
    }

    private fun scheduleRetry(state: String, detail: LinkStatus) {
        if (!running) return
        if (state == "revoked") {
            running = false
            setStatus(detail.copy(state = "revoked", since = System.currentTimeMillis()))
            return
        }
        val base = minOf(backoff.maxMs, backoff.minMs * (1L shl minOf(attempt, 20)))
        val delay = Math.round(base * (0.75 + Math.random() * 0.5))
        attempt++
        setStatus(detail.copy(state = state, since = System.currentTimeMillis(), retryAt = System.currentTimeMillis() + delay))
        val gen = generation
        retryTimer = loop.schedule(delay) {
            retryTimer = null
            if (gen == generation && running) connect()
        }
    }

    /** Loop. Starts one attempt; the blocking part runs on a worker, the result comes back to the loop. */
    private fun connect() {
        val gen = ++generation
        val live = { gen == generation && running }
        setStatus(LinkStatus("connecting"))
        val c = creds
        val sock = try {
            RelaySocket(
                PairingCodec.relayWsUrl(c.relayUrl, "/v1/device"),
                mapOf("authorization" to "Bearer ${c.token}", "x-pleiad-host" to c.hostId, "x-pleiad-device" to c.deviceId),
                connectTimeoutMs,
            )
        } catch (e: Exception) {
            log("remote device: ${e.message}")
            scheduleRetry("offline", LinkStatus("offline", reason = "url"))
            return
        }
        socket = sock
        val fail: (String, LinkStatus) -> Unit = { st, detail ->
            sock.close(1000)
            if (socket === sock) socket = null
            if (live()) scheduleRetry(st, detail)
        }
        workers.execute {
            val o = sock.awaitOpen()
            if (!o.open) {
                loop.post {
                    if (!live()) { sock.close(); return@post }
                    val closeCode = o.closeCode?.takeIf { it != 1006 }
                    val detail = when {
                        closeCode != null -> LinkStatus("", closeCode = closeCode)
                        o.status != null -> LinkStatus("", httpStatus = o.status)
                        else -> LinkStatus("", reason = o.error)
                    }
                    fail(if (closeCode != null) LinkRules.classifyClose(closeCode) else "offline", detail)
                }
                return@execute
            }
            var transport: Transport? = null
            var failure: Pair<String, LinkStatus>? = null
            try {
                val hs = Handshake(Pattern.IK, true, Pleiad.prologueFor(c.hostId), keyPair, remoteStatic = c.hostPublicKey)
                val payload = JSONObject().put("proto", 1).put("name", name).put("app", app)
                sock.send(hs.writeMessage(payload.toString().toByteArray(Charsets.UTF_8)))
                hs.readMessage(sock.next(connectTimeoutMs))
                transport = hs.split()
            } catch (e: RelayClosed) {
                failure = LinkRules.classifyClose(e.code) to LinkStatus("", closeCode = e.code)
            } catch (e: RelayTimeout) {
                failure = "host-offline" to LinkStatus("", reason = "timeout")
            } catch (e: Exception) {
                // The host's key does not match: treat like revoked (possible impersonation), don't retry
                log("remote device: handshake failed: ${e.message}")
                failure = "revoked" to LinkStatus("", reason = "handshake")
            }
            loop.post {
                if (!live()) { sock.close(); return@post }
                val f = failure
                if (f != null) return@post fail(f.first, f.second)
                attach(sock, transport!!, gen, live, fail)
            }
        }
    }

    /** Loop. The handshake is done: build the channel, attach listeners, THEN drain queued messages, then HELLO. */
    private fun attach(sock: RelaySocket, transport: Transport, gen: Int, live: () -> Boolean, fail: (String, LinkStatus) -> Unit) {
        val hello = JSONObject().put("app", app).put("shell", shell)
        val send: (ByteArray) -> Unit = { b -> if (!sock.send(b)) throw IllegalStateException("relay socket refused the message") }
        val buffered = { sock.bufferedAmount() }
        val ch = channelFactory?.invoke(send, transport, hello, buffered)
            ?: Channel(loop, "device", send, transport, hello, bufferedAmount = buffered)
        var goaway: String? = null
        var settled = false
        var helloDone = false
        var helloTimer: Cancellable? = null

        fun onHelloResult(h: JSONObject?, timedOut: Boolean) {
            if (helloDone) return
            helloDone = true
            helloTimer?.cancel()
            if (!live()) { ch.close(); sock.close(); return }
            if (h == null) {
                ch.close()
                // Wait up to 1 s for the relay's close code to classify the failure
                val finish = { closed: RelayClosed? ->
                    loop.post {
                        if (!live()) return@post
                        val cc = closed?.code?.takeIf { it != 1006 }
                        val state = if (timedOut && cc == null) "host-offline" else LinkRules.classifyClose(cc, goaway)
                        fail(state, LinkStatus("", closeCode = cc, goaway = goaway, reason = if (timedOut) "timeout" else null))
                    }
                }
                var answered = false
                val t = loop.schedule(1000) { if (!answered) { answered = true; finish(sock.closed) } }
                sock.onClose { cl -> loop.post { if (!answered) { answered = true; t.cancel(); finish(cl) } } }
                return
            }
            settled = true
            channel = ch
            hostName = (h.opt("hostName") as? String) ?: ""
            connectedAt = System.currentTimeMillis()
            stableTimer?.cancel()
            stableTimer = loop.schedule(backoff.stableMs) { if (channel === ch) attempt = 0 }
            setStatus(LinkStatus("connected"))
            val closedAlready = sock.closed
            if (closedAlready != null && channel === ch) {
                channel = null
                if (live() && state == "connected") scheduleRetry(LinkRules.classifyClose(closedAlready.code.takeIf { it != 1006 }, goaway), LinkStatus("", closeCode = closedAlready.code))
            }
        }

        ch.listener = object : ChannelListener {
            override fun onHello(hello: JSONObject) = onHelloResult(hello, false)
            override fun onGoaway(code: String, reason: String) { goaway = code }
            override fun onClose(err: ChannelError?) {
                sock.close(1000)
                if (!helloDone) onHelloResult(null, false)
            }
        }
        helloTimer = loop.schedule(connectTimeoutMs) { onHelloResult(null, true) }
        sock.onClose { closed ->
            loop.post {
                ch.close(ChannelError("transport", "relay connection lost"))
                if (settled) {
                    if (channel === ch) channel = null
                    stableTimer?.cancel()
                    if (socket === sock) socket = null
                    if (!live()) return@post
                    if (state != "connected") return@post   // handled already (e.g. closed right after HELLO)
                    val cc = closed.code.takeIf { it != 1006 }
                    scheduleRetry(LinkRules.classifyClose(cc, goaway), LinkStatus("", closeCode = cc, goaway = goaway))
                }
            }
        }
        // Listeners are attached: now hand over queued messages (HELLO may already be among them) and later ones, in order.
        sock.drainTo { b -> loop.post { ch.receive(b) } }
        ch.start()
    }
}

File: core/remote/device-link.mjs
// 端末から中継を通ってホストへつなぐ 1 本の線（docs/remote.md §3.2・§4.4・§7.4）。
// デスクトップ版の端末内プロキシ（device-proxy.mjs）が使う。モバイルは同じ流れを Swift / Kotlin で持つ。
//
//   const link = new DeviceLink({ creds, keyPair, app: '0.1.0', name: 'my-laptop' });
//   link.on('status', s => ...);          // { state, closeCode?, reason?, since, retryAt, hostName }
//   link.start();
//   const ch = await link.ready(10_000);  // つながり次第 Channel。ホストが居ない・取り消し済みなら即座に reject
//   link.stop();
//
// 状態（state）:
//   'connecting'    中継へつなぎ、ハンドシェイクと HELLO を待っている
//   'connected'     チャネルが使える
//   'offline'       中継につながらない・通信が切れた（張り直しを待っている）
//   'host-offline'  中継は答えたがホストが居ない（4404・4408・ホストの GOAWAY shutdown）。張り直しを待っている
//   'revoked'       この端末は取り消された・資格が合わない（4401・GOAWAY revoked）。張り直さない
//   'stopped'       stop() した
//
// 張り直しは 0.5 秒から 30 秒まで倍々（±25% の揺らぎ）。10 秒つながり続けたら初めからに戻す。
// ストリームは持ち越さない（§4.4）。切れたらチャネルが全ストリームを捨て、プロキシが 502 / 1006 にする。
import { EventEmitter } from 'node:events';
import WebSocket from 'ws';
import { Handshake, prologueFor } from './noise.mjs';
import { Channel, ChannelError } from './channel.mjs';
import { relayWsUrl } from './pairing.mjs';
import { t } from '../i18n.mjs';

/** 中継への WebSocket の 1 メッセージの上限（中継の 66,000 に余裕を持たせる）。 */
const RELAY_MAX_PAYLOAD = 70_000;
export const LINK_BACKOFF = Object.freeze({ minMs: 500, maxMs: 30_000, stableMs: 10_000 });

/** 中継の close code とホストの GOAWAY から状態を決める。 */
export function classifyClose({ closeCode = null, goaway = null } = {}) {
  if (goaway === 'revoked' || closeCode === 4401) return 'revoked';
  if (goaway === 'shutdown' || closeCode === 4404 || closeCode === 4408) return 'host-offline';
  return 'offline';
}

/**
 * 中継の口へ WebSocket を張る。届いたメッセージは順に溜め、next() で 1 通ずつ取るか drainTo() で流し込む。
 * ハンドシェイクのメッセージ 2 と HELLO が同じ読み出しで届いても取りこぼさない。
 * 中継の close code は closed（{ code, reason }）に入る。
 */
export function openRelaySocket(url, headers, { openTimeoutMs = 10_000, WebSocketImpl = WebSocket } = {}) {
  const ws = new WebSocketImpl(url, { headers, perMessageDeflate: false, maxPayload: RELAY_MAX_PAYLOAD, followRedirects: false, handshakeTimeout: openTimeoutMs });
  let isClosed = null;
  const closed = new Promise(res => ws.on('close', (code, reason) => {
    isClosed = { code, reason: reason?.toString('utf8') ?? '' };
    for (const w of waiters.splice(0)) w(null, isClosed);
    res(isClosed);
  }));
  ws.on('error', () => {});
  const opened = new Promise(res => {
    const timer = setTimeout(() => res({ error: 'open timeout' }), openTimeoutMs + 1000);
    timer.unref?.();
    ws.once('open', () => { clearTimeout(timer); res({ open: true }); });
    ws.once('unexpected-response', (req, r) => { clearTimeout(timer); res({ status: r.statusCode }); r.resume(); req.destroy(); });
    ws.once('error', e => { clearTimeout(timer); res({ error: e.message }); });
    ws.once('close', code => { clearTimeout(timer); res({ closeCode: code }); });
  });
  const inbox = [];
  const waiters = [];
  let sink = null;
  ws.on('message', data => {
    const buf = Buffer.isBuffer(data) ? data : Buffer.concat(Array.isArray(data) ? data : [Buffer.from(data)]);
    if (sink) return sink(buf);
    const w = waiters.shift();
    if (w) w(buf); else inbox.push(buf);
  });
  /** 次の 1 通。閉じられたら reject（closeCode 付き）、ms を過ぎても reject。 */
  const next = (ms = 10_000) => new Promise((resolve, reject) => {
    if (inbox.length) return resolve(inbox.shift());
    const fail = c => reject(Object.assign(new Error(`relay connection closed (${c.code})`), { closeCode: c.code }));
    if (isClosed) return fail(isClosed);
    const w = (buf, c) => { clearTimeout(timer); if (buf) resolve(buf); else fail(c); };
    const timer = setTimeout(() => {
      const i = waiters.indexOf(w);
      if (i >= 0) waiters.splice(i, 1);
      reject(Object.assign(new Error('no response from host'), { code: 'timeout' }));
    }, ms);
    timer.unref?.();
    waiters.push(w);
  });
  /** 以後のメッセージを fn へ（溜まっていた分から順に、同期的に）。聞き手を付け終えてから呼ぶこと。 */
  const drainTo = fn => { sink = fn; for (const b of inbox.splice(0)) fn(b); };
  return { ws, closed, opened, next, drainTo };
}

/** 閉じる。閉じ終わらなければ 2 秒で断ち切る。 */
function closeSocket(ws, code = 1000) {
  if (ws.readyState === WebSocket.CLOSED) return;
  try { ws.readyState === WebSocket.CONNECTING ? ws.terminate() : ws.close(code); } catch { ws.terminate(); }
  const timer = setTimeout(() => { try { ws.terminate(); } catch {} }, 2000);
  timer.unref?.();
}

function errorFor(state, detail) {
  const e = new Error(state === 'revoked' ? t('remote.device.revoked')
    : state === 'host-offline' ? t('remote.device.hostOffline')
      : state === 'stopped' ? t('remote.device.stopped') : t('remote.device.offline'));
  e.state = state;
  if (detail?.closeCode != null) e.closeCode = detail.closeCode;
  return e;
}

export class DeviceLink extends EventEmitter {
  /**
   * creds: { hostId, hostPublicKey (Buffer), relayUrl, deviceId, token }
   * keyPair: 端末の静的鍵 { publicKey, privateKey }
   * backoff: { minMs, maxMs, stableMs }（試験で縮める）
   * connectTimeoutMs: 中継への接続・ハンドシェイク・HELLO のそれぞれの待ちの上限
   */
  constructor({ creds, keyPair, app = '', name = '', shell = 'desktop', backoff = {}, connectTimeoutMs = 15_000, channelOptions = {}, log = () => {} }) {
    super();
    this.creds = creds;
    this.keyPair = keyPair;
    this.app = String(app);
    this.name = String(name);
    this.shell = shell;
    this.backoff = { ...LINK_BACKOFF, ...backoff };
    this.connectTimeoutMs = connectTimeoutMs;
    this.channelOptions = channelOptions;
    this.log = log;
    this.channel = null;
    this.hostName = '';
    this.attempt = 0;
    this.running = false;
    this.generation = 0;
    this.retryTimer = null;
    this.stableTimer = null;
    this.socket = null;
    this._status = { state: 'stopped', since: new Date().toISOString(), retryAt: null };
  }

  get status() { return { ...this._status, hostName: this.hostName }; }
  get state() { return this._status.state; }

  #setStatus(state, extra = {}) {
    this._status = { state, since: new Date().toISOString(), retryAt: null, ...extra };
    this.emit('status', this.status);
  }

  /** つなぎ始める（つながっていれば何もしない）。 */
  start() {
    if (this.running) return;
    this.running = true;
    this.attempt = 0;
    this.#connect();
  }

  /** 止める。張り直しもしない。 */
  stop() {
    if (!this.running && this.state === 'stopped') return;
    this.running = false;
    this.generation++;
    clearTimeout(this.retryTimer);
    clearTimeout(this.stableTimer);
    this.retryTimer = null;
    const ch = this.channel;
    this.channel = null;
    ch?.close(new ChannelError('closed', 'link stopped'));
    if (this.socket) closeSocket(this.socket, 1000);
    this.socket = null;
    this.#setStatus('stopped');
  }

  /** 待たずに今すぐつなぎ直す（「再試行」、取り消し後の再ペアリングで資格が変わったときも）。 */
  retryNow(creds) {
    if (creds) this.creds = creds;
    if (!this.running) { this.start(); return; }
    if (this.state === 'connecting' || this.state === 'connected') { if (!creds) return; }
    this.generation++;
    clearTimeout(this.retryTimer);
    this.retryTimer = null;
    const ch = this.channel;
    this.channel = null;
    ch?.close(new ChannelError('closed', 'reconnecting'));
    if (this.socket) closeSocket(this.socket, 1000);
    this.socket = null;
    this.attempt = 0;
    this.#connect();
  }

  /**
   * 使えるチャネル。つながっていればすぐ、つないでいる最中なら結果を待つ（最大 ms）。
   * 張り直しを待っている・取り消し済み・止めたなら、待たずに reject（err.state に状態）。
   */
  ready(ms = 10_000) {
    if (this.state === 'connected' && this.channel && !this.channel.closed) return Promise.resolve(this.channel);
    if (this.state !== 'connecting') return Promise.reject(errorFor(this.state, this._status));
    return new Promise((resolve, reject) => {
      const onStatus = s => {
        if (s.state === 'connecting') return;
        cleanup();
        if (s.state === 'connected' && this.channel) resolve(this.channel);
        else reject(errorFor(s.state, s));
      };
      const timer = setTimeout(() => { cleanup(); reject(errorFor('offline')); }, ms);
      timer.unref?.();
      const cleanup = () => { clearTimeout(timer); this.off('status', onStatus); };
      this.on('status', onStatus);
    });
  }

  #scheduleRetry(state, detail) {
    if (!this.running) return;
    if (state === 'revoked') {
      this.running = false;
      this.#setStatus('revoked', detail);
      return;
    }
    const base = Math.min(this.backoff.maxMs, this.backoff.minMs * 2 ** this.attempt);
    const delay = Math.round(base * (0.75 + Math.random() * 0.5));
    this.attempt++;
    const retryAt = new Date(Date.now() + delay).toISOString();
    this.#setStatus(state, { ...detail, retryAt });
    const gen = this.generation;
    this.retryTimer = setTimeout(() => { this.retryTimer = null; if (gen === this.generation && this.running) this.#connect(); }, delay);
    this.retryTimer.unref?.();
  }

  async #connect() {
    const gen = ++this.generation;
    const live = () => gen === this.generation && this.running;
    this.#setStatus('connecting');
    const { creds } = this;
    let sock;
    try {
      sock = openRelaySocket(relayWsUrl(creds.relayUrl, '/v1/device'), {
        authorization: `Bearer ${creds.token}`, 'x-pleiad-host': creds.hostId, 'x-pleiad-device': creds.deviceId,
      }, { openTimeoutMs: this.connectTimeoutMs });
    } catch (e) {
      this.log(`remote device: ${e.message}`);
      return this.#scheduleRetry('offline', { reason: 'url' });
    }
    this.socket = sock.ws;
    const fail = (state, detail = {}) => {
      closeSocket(sock.ws, 1000);
      if (this.socket === sock.ws) this.socket = null;
      if (live()) this.#scheduleRetry(state, detail);
    };

    const o = await sock.opened;
    if (!live()) return closeSocket(sock.ws);
    if (!o.open) {
      // 中継の Upgrade 前の拒否（429 など）・名前解決やつながりの失敗・開いた直後に閉じられた
      const detail = o.closeCode != null ? { closeCode: o.closeCode } : o.status ? { httpStatus: o.status } : { reason: o.error };
      return fail(o.closeCode != null ? classifyClose({ closeCode: o.closeCode }) : 'offline', detail);
    }

    let transport;
    try {
      const hs = new Handshake({ pattern: 'IK', initiator: true, prologue: prologueFor(creds.hostId), staticKey: this.keyPair, remoteStatic: creds.hostPublicKey });
      sock.ws.send(hs.writeMessage(Buffer.from(JSON.stringify({ proto: 1, name: this.name, app: this.app }))));
      hs.readMessage(await sock.next(this.connectTimeoutMs));
      transport = hs.split();
    } catch (e) {
      if (!live()) return closeSocket(sock.ws);
      // 閉じられた（4401 取り消し・4404 ホストが居ない など）か、ホストの鍵が合わない（なりすましの疑い = 取り消しと同じく張り直さない）
      if (e.closeCode != null) return fail(classifyClose({ closeCode: e.closeCode }), { closeCode: e.closeCode });
      if (e.code === 'timeout') return fail('host-offline', { reason: 'timeout' });
      this.log(`remote device: handshake failed: ${e.message}`);
      return fail('revoked', { reason: 'handshake' });
    }

    const ch = new Channel({
      role: 'device', transport, send: b => sock.ws.send(b), hello: { app: this.app, shell: this.shell },
      bufferedAmount: () => sock.ws.bufferedAmount, ...this.channelOptions,
    });
    let goaway = null;
    let settled = false;
    const hello = new Promise(resolve => {
      // 聞き手を先に付けてから溜まった分を流す（後から付けると、drainTo の中で同期的に処理された HELLO を取りこぼす）
      ch.once('hello', h => resolve({ hello: h }));
      ch.once('close', () => resolve({ closed: true }));
      const timer = setTimeout(() => resolve({ timeout: true }), this.connectTimeoutMs);
      timer.unref?.();
      ch.once('hello', () => clearTimeout(timer));
      ch.once('close', () => clearTimeout(timer));
    });
    ch.on('goaway', g => { goaway = String(g?.code ?? ''); });
    ch.on('close', () => closeSocket(sock.ws, 1000));
    sock.ws.on('close', () => ch.close(new ChannelError('transport', 'relay connection lost')));
    sock.closed.then(c => {
      if (settled) {
        // つながっていたチャネルが切れた
        if (this.channel === ch) this.channel = null;
        clearTimeout(this.stableTimer);
        if (this.socket === sock.ws) this.socket = null;
        if (!live()) return;
        const detail = { closeCode: c.code, ...(goaway ? { goaway } : {}) };
        this.#scheduleRetry(classifyClose({ closeCode: c.code, goaway }), detail);
      }
    });
    sock.drainTo(b => ch.receive(b));
    ch.start();
    const h = await hello;
    if (!live()) { ch.close(); return closeSocket(sock.ws); }
    if (!h.hello) {
      ch.close();
      const c = await Promise.race([sock.closed, new Promise(r => { const timer = setTimeout(r, 1000, null); timer.unref?.(); })]);
      if (!live()) return;
      const closeCode = c?.code ?? null;
      return fail(h.timeout && closeCode == null ? 'host-offline' : classifyClose({ closeCode, goaway }), { ...(closeCode != null ? { closeCode } : {}), ...(goaway ? { goaway } : {}), ...(h.timeout ? { reason: 'timeout' } : {}) });
    }
    settled = true;
    this.channel = ch;
    this.hostName = typeof h.hello.hostName === 'string' ? h.hello.hostName : '';
    this.stableTimer = setTimeout(() => { if (this.channel === ch) this.attempt = 0; }, this.backoff.stableMs);
    this.stableTimer.unref?.();
    this.emit('channel', ch);
    this.#setStatus('connected');
    // 既に切れていた（HELLO の直後に閉じられた）なら、上の sock.closed が張り直しを受け持つ
    if (sock.ws.readyState === WebSocket.CLOSED && this.channel === ch) {
      this.channel = null;
      const c = await sock.closed;
      if (live() && this.state === 'connected') this.#scheduleRetry(classifyClose({ closeCode: c.code, goaway }), { closeCode: c.code });
    }
  }
}
